src/mpid/ch4/netmod/ofi/ofi_send.h

```diff
@@ -384,6 +384,56 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
     return mpi_errno;
 }
 
+#ifdef MPL_HAVE_GPU
+static inline int MPIDI_NM_do_send(const void *buf, MPI_Aint count,
+                                   MPI_Datatype datatype, int rank, int tag,
+                                   MPIR_Comm * comm, int context_offset,
+                                   MPIDI_av_entry_t * addr, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_NM_DO_SEND);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_NM_DO_SEND);
+
+    bool is_staged = false;
+    MPL_pointer_type_t type;
+    MPL_gpu_query_pointer_type(buf, &type);
+    if (!MPIDI_OFI_global.can_handle_gpu_mem &&
+        MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map,
+                          (uint64_t) buf) != MPIDIU_MAP_NOT_FOUND && type == MPL_GPU_POINTER_DEV) {
```

Comment:
I think the condition should be 
`MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map,(uint64_t) buf) == MPIDIU_MAP_NOT_FOUND`

`NOT_FOUND` means we need to allocate stage buffer for gpu pointer, move data and return its pointer.

Otherwise it means currently there is going-on ops on this gpu pointer.

Same for both send and recv cases.

Reply:
A second thought, we don't need this check `MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf) != MPIDIU_MAP_NOT_FOUND`

Always fetch a CPU buffer for GPU buffer.

MPIDIU_map is used to remember fetched CPU buffer, and to release it correctly in request completion.

Reply:
You are right this design was when we were falling back to the generic layer and then back again into the netmod. This check should be removed now that we are handling everything inside OFI

Reply:
I will do that in my PR.

Reply:
If it's not too much work would it be possible for you to keep working on this. You can overwrite any change you want in this PR (I will assign this to you). Otherwise do you have a PR to replace this already? If so can you update the wiki page with a link to it so that we can keep track of what is going on and close this one?

Reply:
I mean `overwrite` this one when I have updates.

Reply:
After I get local cuda tests passed.